File: PhotoAnalyzer/app/routers/files.py
import os
import uuid
import time
import threading
from concurrent.futures import ThreadPoolExecutor
from pathlib import Path
from datetime import datetime
from fastapi import APIRouter, Query, HTTPException
from fastapi.responses import FileResponse, Response
from ..deps import state
from ..models import FileNode, BrowseResult, ThumbnailJob
from src.config import SUPPORTED_IMAGE_FORMATS, is_image_file

router = APIRouter(prefix="/api", tags=["files"])

THUMB_SIZE = (200, 200)
_thumb_cache: dict[str, str] = {}
_thumb_jobs: dict[str, ThumbnailJob] = {}
_thumb_executor = ThreadPoolExecutor(max_workers=4)

RAW_FORMATS = {".cr2", ".cr3", ".arw", ".dng", ".nef", ".orf", ".rw2", ".pef", ".raf", ".3fr", ".ai", ".eps"}


@router.get("/files", response_model=BrowseResult)
def browse_files(dir_id: str = Query(...), path: str | None = Query(None)):
    entry = state.get_dir(dir_id)
    if not entry:
        from fastapi import HTTPException
        raise HTTPException(404, "目录不存在")

    base = Path(entry.path)
    target = Path(path) if path else base

    try:
        target = target.resolve()
        base = base.resolve()
        if not str(target).startswith(str(base)):
            raise HTTPException(403, "路径超出目录范围")
    except Exception:
        raise HTTPException(400, "无效路径")

    if not target.exists() or not target.is_dir():
        raise HTTPException(404, f"目录不存在: {target}")

    parent_path: str | None = None
    if target != base:
        parent = target.parent
        parent_path = str(parent) if str(parent).startswith(str(base)) else None

    items: list[FileNode] = []
    try:
        stem_map: dict[str, list[Path]] = {}

        for child in sorted(target.iterdir(), key=lambda p: p.name.lower()):
            if child.name.startswith("."):
                continue
            if child.is_dir():
                items.append(FileNode(
                    name=child.name,
                    path=str(child),
                    is_dir=True,
                    size=0,
                    modified="",
                    thumbnail_url=None,
                ))
                continue

            stem = child.stem
            suffix = child.suffix.lower()

            if is_image_file(child) or suffix in RAW_FORMATS:
                if stem not in stem_map:
                    stem_map[stem] = []
                stem_map[stem].append(child)

        for stem, paths in stem_map.items():
            paths.sort(key=lambda p: (0 if p.suffix.lower() == ".jpg" else 1, p.suffix.lower(), p.name.lower()))
            best = paths[0]
            thumb_url = f"/api/thumbnails?path={_encode_path(str(best))}" if best.suffix.lower() in {".jpg", ".jpeg", ".png", ".webp"} else None
            items.append(FileNode(
                name=best.name,
                path=str(best),
                is_dir=False,
                size=best.stat().st_size,
                modified=_format_time(best.stat().st_mtime),
                thumbnail_url=thumb_url,
            ))
    except PermissionError:
        pass

    return BrowseResult(
        current_path=str(target),
        parent_path=parent_path,
        items=items,
    )


@router.get("/thumbnails")
def get_thumbnail(path: str = Query(...), full: bool = Query(False)):
    img_path = Path(path)
    if not img_path.exists() or not is_image_file(img_path):
        raise HTTPException(404, "文件不存在")

    if full:
        return FileResponse(img_path, media_type=f"image/{img_path.suffix.lstrip('.').lower()}")

    cache_key = f"thumb_{path}"
    cached = _thumb_cache.get(cache_key)
    if cached and Path(cached).exists():
        return FileResponse(cached, media_type="image/jpeg")

    thumb_dir = Path(__file__).resolve().parent.parent.parent / "data" / "thumbs"
    thumb_dir.mkdir(parents=True, exist_ok=True)
    thumb_path = thumb_dir / f"{img_path.stem}_{img_path.stat().st_size}.jpg"

    if thumb_path.exists():
        _thumb_cache[cache_key] = str(thumb_path)
        return FileResponse(thumb_path, media_type="image/jpeg")

    return Response(status_code=204)


@router.post("/thumbnails/batch")
def start_thumbnail_batch(paths: list[str]):
    valid_paths = [p for p in paths if Path(p).exists() and is_image_file(Path(p))]
    if not valid_paths:
        raise HTTPException(400, "没有有效的图片路径")

    job_id = str(uuid.uuid4())
    job = ThumbnailJob(
        job_id=job_id,
        status="pending",
        total=len(valid_paths),
        completed=0,
        failed=0,
        created_at=time.strftime("%Y-%m-%dT%H:%M:%S"),
    )
    _thumb_jobs[job_id] = job

    def generate_thumbnails():
        _thumb_jobs[job_id].status = "running"
        for i, p in enumerate(valid_paths):
            if _thumb_jobs[job_id].status == "canceled":
                break
            _thumb_jobs[job_id].current_file = Path(p).name
            _thumb_jobs[job_id].progress = i
            try:
                img_path = Path(p)
                cache_key = f"thumb_{p}"
                thumb_dir = Path(__file__).resolve().parent.parent.parent / "data" / "thumbs"
                thumb_dir.mkdir(parents=True, exist_ok=True)
                thumb_path = thumb_dir / f"{img_path.stem}_{img_path.stat().st_size}.jpg"
                if not thumb_path.exists():
                    from PIL import Image, ImageOps
                    img = Image.open(img_path)
                    img = ImageOps.exif_transpose(img)
                    img.thumbnail(THUMB_SIZE)
                    img = img.convert("RGB")
                    img.save(thumb_path, "JPEG", quality=75)
                _thumb_cache[cache_key] = str(thumb_path)
                _thumb_jobs[job_id].completed += 1
            except Exception:
                _thumb_jobs[job_id].failed += 1

        _thumb_jobs[job_id].status = "completed"
        _thumb_jobs[job_id].finished_at = time.strftime("%Y-%m-%dT%H:%M:%S")

    _thumb_executor.submit(generate_thumbnails)
    return job


@router.get("/thumbnails/batch/{job_id}")
def get_thumbnail_job(job_id: str):
    job = _thumb_jobs.get(job_id)
    if not job:
        raise HTTPException(404, "任务不存在")
    return job


@router.post("/thumbnails/batch/{job_id}/cancel")
def cancel_thumbnail_job(job_id: str):
    job = _thumb_jobs.get(job_id)
    if not job:
        raise HTTPException(404, "任务不存在")
    job.status = "canceled"
    job.finished_at = time.strftime("%Y-%m-%dT%H:%M:%S")
    return job

def _encode_path(p: str) -> str:
    import urllib.parse
    return urllib.parse.quote(p, safe="")


@router.delete("/files")
def delete_file(path: str = Query(...)):
    p = Path(path)
    if not p.exists():
        raise HTTPException(404, "文件不存在")
    if not p.is_file():
        raise HTTPException(400, "不是文件")
    try:
        p.unlink()
        return {"deleted": [str(p)], "count": 1}
    except Exception as e:
        raise HTTPException(500, f"删除失败: {e}")


@router.get("/files/siblings")
def get_file_siblings(path: str = Query(...)):
    p = Path(path)
    if not p.exists():
        raise HTTPException(404, "文件不存在")
    parent = p.parent
    stem = p.stem
    siblings = []
    for f in parent.iterdir():
        if f.is_file() and f.stem == stem and f != p:
            siblings.append(str(f))
    return {"siblings": siblings, "count": len(siblings)}


RAW_EXTENSIONS = {".cr2", ".arw", ".dng", ".nef", ".orf", ".rw2", ".pef", ".srw", ".raf"}


@router.get("/files/orphaned-raws")
def get_orphaned_raws(dir_id: str = Query(...)):
    from ..deps import state
    entry = state.get_dir(dir_id)
    if not entry:
        raise HTTPException(404, "目录不存在")
    target = Path(entry.path)
    orphaned = []
    for f in target.rglob("*"):
        if not f.is_file():
            continue
        ext = f.suffix.lower()
        if ext not in RAW_EXTENSIONS:
            continue
        stem = f.stem
        jpg_path = f.parent / f"{stem}.jpg"
        jpg_path_upper = f.parent / f"{stem}.JPG"
        if not jpg_path.exists() and not jpg_path_upper.exists():
            orphaned.append(str(f))
    return {"orphaned": orphaned, "count": len(orphaned)}


@router.delete("/files/orphaned-raws")
def delete_orphaned_raws(dir_id: str = Query(...)):
    from ..deps import state
    entry = state.get_dir(dir_id)
    if not entry:
        raise HTTPException(404, "目录不存在")
    target = Path(entry.path)
    deleted = []
    not_found = []
    for f in target.rglob("*"):
        if not f.is_file():
            continue
        ext = f.suffix.lower()
        if ext not in RAW_EXTENSIONS:
            continue
        stem = f.stem
        jpg_path = f.parent / f"{stem}.jpg"
        jpg_path_upper = f.parent / f"{stem}.JPG"
        if not jpg_path.exists() and not jpg_path_upper.exists():
            if f.exists():
                f.unlink()
                deleted.append(str(f))
            else:
                not_found.append(str(f))
    return {"deleted": deleted, "not_found": not_found, "count": len(deleted)}


def _format_time(mtime: float) -> str:
    try:
        return datetime.fromtimestamp(mtime).strftime("%Y-%m-%d %H:%M:%S")
    except Exception:
        return ""

File: PhotoAnalyzer/rust/src/handlers/files.rs
use axum::{extract::{Query, State}, Json};
use std::collections::HashMap;
use std::sync::Arc;
use std::path::Path;

use crate::models::{BrowseResult, FileNode};
use crate::services::AppState;

const IMAGE_EXTENSIONS: &[&str] = &["jpg", "jpeg", "png", "gif", "bmp", "webp", "tiff"];
const RAW_EXTENSIONS: &[&str] = &["cr2", "cr3", "arw", "dng", "nef", "orf", "rw2", "pef", "srw", "raf", "3fr", "ai", "eps"];

pub async fn browse_files(
    State(state): State<Arc<AppState>>,
    Query(params): Query<BrowseQuery>,
) -> Result<Json<BrowseResult>, (axum::http::StatusCode, &'static str)> {
    let dir = state.get_dir(&params.dir_id).ok_or((axum::http::StatusCode::NOT_FOUND, "directory not found"))?;

    let target = params.path.as_ref()
        .map(Path::new)
        .unwrap_or_else(|| Path::new(&dir.path))
        .to_path_buf();

    if !target.exists() {
        return Err((axum::http::StatusCode::NOT_FOUND, "path not found"));
    }

    let base = Path::new(&dir.path).to_path_buf();
    let parent_path = if target != base {
        target.parent().map(|p| p.to_string_lossy().to_string())
    } else {
        None
    };

    let mut items = Vec::new();
    let mut stem_map: HashMap<String, Vec<std::path::PathBuf>> = HashMap::new();

    let read_dir = match std::fs::read_dir(&target) {
        Ok(rd) => rd,
        Err(_) => return Err((axum::http::StatusCode::INTERNAL_SERVER_ERROR, "failed to read directory")),
    };

    for entry in read_dir.filter_map(|e| e.ok()) {
        let path = entry.path();
        let name = path.file_name()
            .map(|s| s.to_string_lossy().to_string())
            .unwrap_or_default();

        if name.starts_with('.') {
            continue;
        }

        if path.is_dir() {
            let modified = path.metadata()
                .ok()
                .and_then(|m| m.modified().ok())
                .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                .map(|d| {
                    chrono::DateTime::from_timestamp(d.as_secs() as i64, 0)
                        .unwrap_or_default()
                        .format("%Y-%m-%d %H:%M:%S")
                        .to_string()
                })
                .unwrap_or_default();

            items.push(FileNode {
                name,
                path: path.to_string_lossy().to_string(),
                is_dir: true,
                size: 0,
                modified,
                thumbnail_url: None,
            });
            continue;
        }

        let ext = path.extension()
            .map(|s| s.to_string_lossy().to_lowercase())
            .unwrap_or_default();

        if is_image_file(&ext) || RAW_EXTENSIONS.contains(&ext.as_str()) {
            let stem = path.file_stem()
                .map(|s| s.to_string_lossy().to_string())
                .unwrap_or_default();
            stem_map.entry(stem).or_default().push(path);
        }
    }

    let mut stem_keys: Vec<_> = stem_map.keys().cloned().collect();
    stem_keys.sort_by(|a, b| a.to_lowercase().cmp(&b.to_lowercase()));

    for stem in stem_keys {
        let mut paths = stem_map.get(&stem).unwrap().clone();
        paths.sort_by(|a, b| {
            let a_ext = a.extension().map(|s| s.to_string_lossy().to_lowercase()).unwrap_or_default();
            let b_ext = b.extension().map(|s| s.to_string_lossy().to_lowercase()).unwrap_or_default();
            let a_is_jpg = a_ext == "jpg";
            let b_is_jpg = b_ext == "jpg";
            if a_is_jpg != b_is_jpg {
                std::cmp::Ordering::Less
            } else {
                a_ext.cmp(&b_ext)
            }
        });

        let best = &paths[0];
        let best_ext = best.extension()
            .map(|s| s.to_string_lossy().to_lowercase())
            .unwrap_or_default();
        let thumbnail_url = if best_ext == "jpg" || best_ext == "jpeg" || best_ext == "png" || best_ext == "webp" {
            Some(format!("/api/thumbnails?path={}", encode_uri(&best.to_string_lossy())))
        } else {
            None
        };

        let metadata = best.metadata().ok();
        let size = metadata.as_ref().map(|m| m.len()).unwrap_or(0);
        let modified = metadata
            .and_then(|m| m.modified().ok())
            .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
            .map(|d| {
                chrono::DateTime::from_timestamp(d.as_secs() as i64, 0)
                    .unwrap_or_default()
                    .format("%Y-%m-%d %H:%M:%S")
                    .to_string()
            })
            .unwrap_or_default();

        items.push(FileNode {
            name: best.file_name().map(|s| s.to_string_lossy().to_string()).unwrap_or_default(),
            path: best.to_string_lossy().to_string(),
            is_dir: false,
            size,
            modified,
            thumbnail_url,
        });
    }

    Ok(Json(BrowseResult {
        current_path: target.to_string_lossy().to_string(),
        parent_path,
        items,
    }))
}

#[derive(serde::Deserialize)]
pub struct BrowseQuery {
    dir_id: String,
    path: Option<String>,
}

pub async fn delete_file(
    Query(params): Query<DeleteQuery>,
) -> Result<Json<DeleteResponse>, (axum::http::StatusCode, &'static str)> {
    let path = Path::new(&params.path);

    if !path.exists() {
        return Err((axum::http::StatusCode::NOT_FOUND, "file not found"));
    }

    if !path.is_file() {
        return Err((axum::http::StatusCode::BAD_REQUEST, "not a file"));
    }

    std::fs::remove_file(path).map_err(|_| (axum::http::StatusCode::INTERNAL_SERVER_ERROR, "delete failed"))?;

    Ok(Json(DeleteResponse {
        deleted: params.path,
        count: 1,
    }))
}

#[derive(serde::Deserialize)]
pub struct DeleteQuery {
    path: String,
}

#[derive(serde::Serialize)]
pub struct DeleteResponse {
    deleted: String,
    count: usize,
}

pub async fn get_siblings(
    Query(params): Query<DeleteQuery>,
) -> Result<Json<SiblingsResponse>, (axum::http::StatusCode, &'static str)> {
    let p = Path::new(&params.path);

    if !p.exists() {
        return Err((axum::http::StatusCode::NOT_FOUND, "file not found"));
    }

    let parent = p.parent().ok_or((axum::http::StatusCode::NOT_FOUND, "no parent"))?;
    let stem = p.file_stem()
        .map(|s| s.to_string_lossy().to_string())
        .unwrap_or_default();

    let mut siblings = Vec::new();

    let read_dir = match std::fs::read_dir(parent) {
        Ok(rd) => rd,
        Err(_) => return Ok(Json(SiblingsResponse { siblings: vec![], count: 0 })),
    };

    for entry in read_dir.filter_map(|e| e.ok()) {
        let path = entry.path();
        if path == p {
            continue;
        }
        if path.is_file() {
            if let Some(entry_stem) = path.file_stem().map(|s| s.to_string_lossy().to_string()) {
                if entry_stem == stem {
                    siblings.push(path.to_string_lossy().to_string());
                }
            }
        }
    }

    let count = siblings.len();
    Ok(Json(SiblingsResponse {
        siblings,
        count,
    }))
}

#[derive(serde::Serialize)]
pub struct SiblingsResponse {
    siblings: Vec<String>,
    count: usize,
}

pub async fn get_orphaned_raws(
    State(state): State<Arc<AppState>>,
    Query(params): Query<OrphanedQuery>,
) -> Result<Json<OrphanedResponse>, (axum::http::StatusCode, &'static str)> {
    let entry = state.get_dir(&params.dir_id).ok_or((axum::http::StatusCode::NOT_FOUND, "directory not found"))?;

    let target = Path::new(&entry.path);
    let mut orphaned = Vec::new();

    for entry in walkdir::WalkDir::new(target)
        .into_iter()
        .filter_map(|e| e.ok())
    {
        let path = entry.path();
        if path.is_file() {
            if let Some(ext) = path.extension().map(|s| s.to_string_lossy().to_lowercase()) {
                if RAW_EXTENSIONS.contains(&ext.as_str()) {
                    let jpg_path = path.with_extension("jpg");
                    let jpg_path_upper = path.with_extension("JPG");

                    if !jpg_path.exists() && !jpg_path_upper.exists() {
                        orphaned.push(path.to_string_lossy().to_string());
                    }
                }
            }
        }
    }

    let count = orphaned.len();
    Ok(Json(OrphanedResponse {
        orphaned,
        count,
    }))
}

#[derive(serde::Deserialize)]
pub struct OrphanedQuery {
    dir_id: String,
}

#[derive(serde::Serialize)]
pub struct OrphanedResponse {
    orphaned: Vec<String>,
    count: usize,
}

pub async fn delete_orphaned_raws(
    State(state): State<Arc<AppState>>,
    Query(params): Query<OrphanedQuery>,
) -> Result<Json<DeleteOrphanedResponse>, (axum::http::StatusCode, &'static str)> {
    let entry = state.get_dir(&params.dir_id).ok_or((axum::http::StatusCode::NOT_FOUND, "directory not found"))?;

    let target = Path::new(&entry.path);
    let mut deleted = Vec::new();
    let mut not_found = Vec::new();

    for entry in walkdir::WalkDir::new(target)
        .into_iter()
        .filter_map(|e| e.ok())
    {
        let path = entry.path();
        if path.is_file() {
            if let Some(ext) = path.extension().map(|s| s.to_string_lossy().to_lowercase()) {
                if RAW_EXTENSIONS.contains(&ext.as_str()) {
                    let jpg_path = path.with_extension("jpg");
                    let jpg_path_upper = path.with_extension("JPG");

                    if !jpg_path.exists() && !jpg_path_upper.exists() {
                        if path.exists() {
                            let _ = std::fs::remove_file(path);
                            deleted.push(path.to_string_lossy().to_string());
                        } else {
                            not_found.push(path.to_string_lossy().to_string());
                        }
                    }
                }
            }
        }
    }

    let count = deleted.len();
    Ok(Json(DeleteOrphanedResponse {
        deleted,
        not_found,
        count,
    }))
}

#[derive(serde::Serialize)]
pub struct DeleteOrphanedResponse {
    deleted: Vec<String>,
    not_found: Vec<String>,
    count: usize,
}

fn is_image_file(ext: &str) -> bool {
    IMAGE_EXTENSIONS.contains(&ext)
}

fn encode_uri(input: &str) -> String {
    let mut result = String::new();
    for c in input.chars() {
        match c {
            'A'..='Z' | 'a'..='z' | '0'..='9' | '-' | '_' | '.' | '~' => result.push(c),
            _ => {
                for byte in c.to_string().as_bytes() {
                    result.push_str(&format!("%{:02X}", byte));
                }
            }
        }
    }
    result
}
